### PR TITLE
week 12 code summary

### PR DESCRIPTION
**Pull request by fg746**
updating scribe notes with code summary from the lab

### CODE REVIEW

File: docs/en/week12/12-3.md [hunk 1]
```diff
@@ -233,3 +233,266 @@ Attention mechanisms allow us to parallelize the operations and greatly accelera
...
+        y = batch.label.to(device)
+
+        out = model(x)
+        acc += (out.argmax(1) == y).cpu().numpy().mean()
+
+    print(f"Eval accuracy: {acc/idx}")
+
+train(train_loader, valid_loader)
+
+```
```
Comment (Atcold) [inline docs/en/week12/12-3.md:498]: Not relevant, plz, remove.

File: docs/en/week12/12-3.md [hunk 2]
```diff
@@ -233,3 +233,266 @@ Attention mechanisms allow us to parallelize the operations and greatly accelera
...
+        x = self.encoder(x)
+
+        x, _ = torch.max(x, dim=1)
+        x = self.dense(x)
+        return x
+
+
+model = TransformerClassifier(num_layers=1, d_model=32, num_heads=2,
+                         conv_hidden_dim=128, input_vocab_size=50002, num_answers=2)
+```
```
Comment (Atcold) [inline docs/en/week12/12-3.md:444]: Shrink down to 3 lines: line 430, 432, and 442.

File: docs/en/week12/12-3.md [hunk 3]
```diff
@@ -233,3 +233,266 @@ Attention mechanisms allow us to parallelize the operations and greatly accelera
...
+            self.enc_layers.append(EncoderLayer(d_model, num_heads, ff_hidden_dim, p))
+
+    def forward(self, x):
+        x = self.embedding(x) # Transform to (batch_size, input_seq_length, d_model)
+
+        for i in range(self.num_layers):
+            x = self.enc_layers[i](x)
+
+        return x  # (batch_size, input_seq_len, d_model)
+```
```
Comment (Atcold) [inline docs/en/week12/12-3.md:411]: Shrink down to 3, 4 lines.

File: docs/en/week12/12-3.md [hunk 4]
```diff
@@ -233,3 +233,266 @@ Attention mechanisms allow us to parallelize the operations and greatly accelera
...
+        out1 = self.layernorm1(x + attn_output)  # (batch_size, input_seq_len, d_model)
+
+        # Feed forward
+        cnn_output = self.cnn(out1)  # (batch_size, input_seq_len, d_model)
+
+        #Second layer norm after adding residual connection
+        out2 = self.layernorm2(out1 + cnn_output)  # (batch_size, input_seq_len, d_model)
+
+        return out2
+```
```
Comment (Atcold) [inline docs/en/week12/12-3.md:370]: Shrink down to 3, 4 lines.

File: docs/en/week12/12-3.md [hunk 5]
```diff
@@ -233,3 +233,266 @@ Attention mechanisms allow us to parallelize the operations and greatly accelera
...
+        self.k1convL1 = nn.Linear(d_model,    hidden_dim)
+        self.k1convL2 = nn.Linear(hidden_dim, d_model)
+        self.activation = nn.ReLU()
+
+    def forward(self, x):
+        x = self.k1convL1(x)
+        x = self.activation(x)
+        x = self.k1convL2(x)
+        return x
+```
```
Comment (Atcold) [inline docs/en/week12/12-3.md:339]: We know how to code a CNN. Plz, remove.

File: docs/en/week12/12-3.md [hunk 6]
```diff
@@ -233,3 +233,266 @@ Attention mechanisms allow us to parallelize the operations and greatly accelera
...
+        H_cat, A = self.scaled_dot_product_attention(Q, K, V)
+
+        # Put all the heads back together by concat
+        H_cat = self.group_heads(H_cat, batch_size)    # (bs, q_length, dim)
+
+        # Final linear layer
+        H = self.W_h(H_cat)          # (bs, q_length, dim)
+
+        return H, A
+```
```
Comment (Atcold) [inline docs/en/week12/12-3.md:317]: Plz, split and describe each method separately.

File: docs/en/week12/12-3.md [hunk 7]
```diff
@@ -255,17 +257,20 @@ class MultiHeadAttention(nn.Module):
...
         self.W_q = nn.Linear(d_xq, d_model, bias=False)
         self.W_k = nn.Linear(d_xk, d_model, bias=False)
         self.W_v = nn.Linear(d_xv, d_model, bias=False)
-
         # Outputs of all sub-layers need to be of dimension d_model
         self.W_h = nn.Linear(d_model, d_model)
 
+```
+
+Initialization of multi-headed attention class. If a d_input is provided, this becomes cross attention. Otherwise, self-attention. The query, key, value setup is constructed as a linear transformation of the input d_model.
```
Comment (Atcold) [inline docs/en/week12/12-3.md:267]:
```suggestion
Initialization of multi-headed attention class. If a `d_input` is provided, this becomes cross attention. Otherwise, self-attention. The query, key, value setup is constructed as a linear transformation of the input `d_model`.
```

File: docs/en/week12/12-3.md [hunk 8]
```diff
@@ -280,92 +285,66 @@ class MultiHeadAttention(nn.Module):
...
-        """
-        Split the last dimension into (heads X depth)
-        Return after transpose to put in shape (batch_size X num_heads X seq_length X d_k)
-        """
+
         return x.view(batch_size, -1, self.num_heads, self.d_k).transpose(1, 2)
+```
+
 
+Split the last dimension into (heads X depth). Return after transpose to put in shape (batch_size X num_heads X seq_length X d_k)
```
Comment (Atcold) [inline docs/en/week12/12-3.md:297]:
```suggestion
Split the last dimension into (`heads` × `depth`). Return after transpose to put in shape (`batch_size` × `num_heads` × `seq_length` × `d_k`)
```

File: docs/en/week12/12-3.md [hunk 9]
```diff
@@ -280,92 +285,66 @@ class MultiHeadAttention(nn.Module):
...
+
+
+```python
     def group_heads(self, x, batch_size):
         """
         Combine the heads again to get (batch_size X seq_length X num_heads X d_k)
         """
```
Comment (Atcold) [inline docs/en/week12/12-3.md]:
```suggestion
```

File: docs/en/week12/12-3.md [hunk 10]
```diff
@@ -280,92 +285,66 @@ class MultiHeadAttention(nn.Module):
...
-        x = self.k1convL2(x)
-        return x
-```
+Although the next block shown in the transformer/encoder's is the Add,Norm, which is a function already built into Pytorch. As such, it is an extremely simple implementation, and does not need it's own class. Next is the 1-D convolution block. Please refer to previous sections for more details.
 
 Now that we have all of our main classes built (or built for us), we now turn to an encoder module.
 
 
 ```python
 class EncoderLayer(nn.Module):
```
Comment (Atcold) [inline docs/en/week12/12-3.md:338]:
```suggestion
class EncoderLayer(nn.Module):
    def __init__(self, d_model, num_heads, conv_hidden_dim, p=0.1):
```

File: docs/en/week12/12-3.md [hunk 11]
```diff
@@ -280,92 +285,66 @@ class MultiHeadAttention(nn.Module):
...
-    def __init__(self, d_model, num_heads, conv_hidden_dim, p=0.1):
-        super().__init__()
-
         self.mha = MultiHeadAttention(d_model, num_heads, p)
-        self.cnn = CNN(d_model, conv_hidden_dim, p)
-
         self.layernorm1 = nn.LayerNorm(normalized_shape=d_model, eps=1e-6)
         self.layernorm2 = nn.LayerNorm(normalized_shape=d_model, eps=1e-6)
```
Comment (Atcold) [inline docs/en/week12/12-3.md:341]:
```suggestion
        self.layernorm2 = nn.LayerNorm(normalized_shape=d_model, eps=1e-6)

    def forward(self, x):
```

File: docs/en/week12/12-3.md [hunk 12]
```diff
@@ -280,92 +285,66 @@ class MultiHeadAttention(nn.Module):
...
-
-    def forward(self, x):
-
-        # Multi-head attention
         attn_output, _ = self.mha(x, x, x)  # (batch_size, input_seq_len, d_model)
-
-        # Layer norm after adding the residual connection
         out1 = self.layernorm1(x + attn_output)  # (batch_size, input_seq_len, d_model)
```
Comment (Atcold) [inline docs/en/week12/12-3.md:343]:
```suggestion
        out1 = self.layernorm1(x + attn_output)  # (batch_size, input_seq_len, d_model)
        cnn_output = self.cnn(out1)  # (batch_size, input_seq_len, d_model)
```

File: docs/en/week12/12-3.md [hunk 13]
```diff
@@ -389,110 +370,36 @@ An entire encoder, with N stacked encoder layers, as well as position embeddings
...
 
 
 ```python
-class TransformerClassifier(nn.Module):
-    def __init__(self, num_layers, d_model, num_heads, conv_hidden_dim, input_vocab_size, num_answers):
-        super().__init__()
-
-        self.encoder = Encoder(num_layers, d_model, num_heads, conv_hidden_dim, input_vocab_size,
-                         maximum_position_encoding=10000)
+        self.encoder = Encoder(num_layers, d_model, num_heads, conv_hidden_dim, input_vocab_size,maximum_position_encoding=10000)
```
Comment (Atcold) [inline docs/en/week12/12-3.md:396]:
```suggestion
class TransformerClassifier(nn.Module):
        self.encoder = Encoder(num_layers, d_model, num_heads, conv_hidden_dim, input_vocab_size, maximum_position_encoding=10000)
```

**Comment by Atcold** [comment]
Thanks!
